packages/store/src/selectors/selector-checks.util.ts

```diff
@@ -1,17 +1,39 @@
 import { ɵgetSelectorMetadata, ɵgetStoreMetadata } from '@ngxs/store/internals';
-
 import { ɵSelectorDef } from './selector-types.util';
 
+function getMissingMetaDataError(
+  selector: ɵSelectorDef<any>,
+  context: { prefix?: string; noun?: string } = {}
+) {
+  const metadata = ɵgetSelectorMetadata(selector) || ɵgetStoreMetadata(selector as any);
+  if (!metadata) {
+    return new Error(
+      `${context.prefix}The value provided as the ${context.noun} is not a valid selector.`
+    );
+  }
+  return null;
+}
+
 export function ensureValidSelector(
   selector: ɵSelectorDef<any>,
   context: { prefix?: string; noun?: string } = {}
 ) {
   const noun = context.noun || 'selector';
   const prefix = context.prefix ? context.prefix + ': ' : '';
   ensureValueProvided(selector, { noun, prefix: context.prefix });
-  const metadata = ɵgetSelectorMetadata(selector) || ɵgetStoreMetadata(selector as any);
-  if (!metadata) {
-    throw new Error(`${prefix}The value provided as the ${noun} is not a valid selector.`);
+  const error = getMissingMetaDataError(selector, { noun, prefix });
+  if (error) {
+    // If we have used this utility within a state class, we may be
+    //  before the @State or @Selector decorators have been applied.
+    //  wait until the next microtask to verify.
+    Promise.resolve().then(() => {
```

Comment:
Since this change may result in breaking unit/e2e tests that implicitly rely on a specific number and sequence of change detections for their assertions to pass:

```suggestion
const isNgZoneEnabled = typeof Zone !== 'undefined';
const runOutsideAngular = isNgZoneEnabled ? (fn) => Zone.root.run(fn) : fn => fn();

    runOutsideAngular(() => {
      Promise.resolve().then(() => {   
```